source/optional/match.d

```diff
@@ -15,38 +15,71 @@ import optional.optional;
         handlers = 2 predicates, one that takes the underlying optional type and another that names nothing
 */
 public template match(handlers...) if (handlers.length == 2) {
-	auto match(T)(inout auto ref Optional!T opt) {
+	auto match(O)(auto ref O opt) {
+		static if (is(O == Optional!T, T)) {
+	        static if (is(typeof(handlers[0](opt.front)))) {
+	            alias someHandler = handlers[0];
+	            alias noHandler = handlers[1];
+				return doMatch!(someHandler, noHandler, O, T)(opt);
+	        } else static if (is(typeof(handlers[0]()))) {
+	            alias someHandler = handlers[1];
+	            alias noHandler = handlers[0];
+				return doMatch!(someHandler, noHandler, O, T)(opt);
+	        } else {
+				// One of these two is causing a compile error.
+				// Let's call them so the compiler can show a proper error warning.
+				failOnCompileError!(handlers[0], T);
+				failOnCompileError!(handlers[1], T);
+			}
+		} else static if (is(typeof(opt.value) == Optional!T, T)) {
+			return opt.valueMatch!handlers;
+		} else {
+			pragma(msg, "Type of: " ~ typeof(opt.value).stringof);
```

Comment:
Stray pragma? Or add into the `static_assert`?

Reply:
Stray pragma indeed. I'll remove this one